File: StackFriday/StackFriday/Stack.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace StackFriday
{
    class Stack<T>
    {
        private Node<T> top;
        public int count { get; private set; }

        /// <summary>
        /// Removes all objects from the Stack.
        /// </summary>
        public void Clear()
        {
            if (top == null)
            {
                return;
            }
            Node<T> next = top.nextNode;
            while (next != null)
            {
                top = null;
                top = next;
                next = next.nextNode;
            }
        }

        /// <summary>
        /// Returns the object at the top of the Stack without removing it.
        /// </summary>
        /// <returns> Returns value on the top. </returns>
        public T Peek()
        {
            return top.value;
        }

        /// <summary>
        /// Returns number of objects in stack.
        /// </summary>
        /// <returns> Returns number of objects in stack. </returns>
        public int Count()
        {
            return count;
        }

        /// <summary>
        /// Removes object at the top of the Stack.
        /// </summary>
        /// <returns> Value of removing object. </returns>
        public T Pop()
        {
            if (!IsEmpty())
            {
                count--;
                Node<T> toRemove = top;
                top = top.nextNode;
                return toRemove.value;
            }
            else
            {
                throw new NullReferenceException();
            }
        }

        /// <summary>
        /// Add new element to stack.
        /// </summary>
        /// <param name="value"> New value of node to insert in stack. </param>
        public void Push(T value)
        {
            Node<T> newNode = new Node<T>();
            newNode.value = value;
            newNode.nextNode = top;
            top = newNode;
            count++;
        }

        /// <summary>
        /// Returns true if stack empty.
        /// </summary>
        /// <returns> Returns true if stack empty. </returns>
        public bool IsEmpty()
        {
            return top == null ? true : false;
        }

    }
}

File: Stack/src/main/java/Stack.java

import java.util.NoSuchElementException;

/**
 * Created by siarhey on 01.12.17.
 */
public class Stack {

  private Element top;
  private int numberOfElements;

  /**
   * Creats an empty stack and sets {@link Stack#numberOfElements} equal 0.
   */
  public Stack() {
    numberOfElements = 0;
  }

  /**
   * Pushes gotten item on the op of the stack.
   * @param gottenItem item we are pushing.
   */
  public void push(Object gottenItem) {
    Element oldTop = top;
    top = new Element ();
    top.setItem(gottenItem);
    top.setPrevious(oldTop);
    numberOfElements++;
  }

  /**
   * Returns the object at the top of this stack without removing it from the stack.
   * If the stack is empty returns -1.
   */
  public Object peek() {
    if(numberOfElements > 0) {
      return top.getItem();
    } else {
      return -1;
    }
  }

  /**
   * Removes the object at the top of this stack and returns that object as the value of this method.
   */
  public Object pop() {
    Object item;
    if (numberOfElements == 0) {
      throw new NoSuchElementException("Stack underflow");
    }
    if (numberOfElements == 1) {
      item = top.getItem();
      top = null;
      numberOfElements--;
      return item;
    }
    top = top.getPrevious();
    numberOfElements--;
    return top.getItem();
  }

  /**
   * Returns true if the stack is empty and false if it is not.
   */
  public boolean isEmpty() {
    return numberOfElements == 0;
  }

  /**
   * Returns number of elements in stack.
   */
  public int getNumberOfElements() {
    return numberOfElements;
  }

  /**
   * Returns the index of item we are are searching for.
   * If there is no such element returns -1.
   * @param item item we are searching.
   */
  public int search(Object item) {
    Element temp = top;
    if (top.getItem() == item) {
      return numberOfElements - 1;
    } else {
      int index = numberOfElements - 2;
      while (temp.getPrevious() != null) {
        temp = temp.getPrevious();
        if (temp.getItem() == item) {
          return index;
        }
        index--;
      }
      return -1;
    }
  }

  public void clear() {
    while (numberOfElements > 0) {
      pop();
    }
  }
}


